File: py/dolarymonedas_sdk.py
# DolarYMonedas SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import DolarYMonedasUtility
from core.spec import DolarYMonedasSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import DolarYMonedasBaseFeature
from features import _make_feature


class DolarYMonedasSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = DolarYMonedasUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return DolarYMonedasUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = DolarYMonedasSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Blue(self, data=None) -> "BlueEntity":
        """Entity factory: client.Blue().list() / client.Blue().load({"id": ...})."""
        from entity.blue_entity import BlueEntity
        return BlueEntity(self, data)


    def Bolsa(self, data=None) -> "BolsaEntity":
        """Entity factory: client.Bolsa().list() / client.Bolsa().load({"id": ...})."""
        from entity.bolsa_entity import BolsaEntity
        return BolsaEntity(self, data)


    def Brl(self, data=None) -> "BrlEntity":
        """Entity factory: client.Brl().list() / client.Brl().load({"id": ...})."""
        from entity.brl_entity import BrlEntity
        return BrlEntity(self, data)


    def Clp(self, data=None) -> "ClpEntity":
        """Entity factory: client.Clp().list() / client.Clp().load({"id": ...})."""
        from entity.clp_entity import ClpEntity
        return ClpEntity(self, data)


    def Contadoconliqui(self, data=None) -> "ContadoconliquiEntity":
        """Entity factory: client.Contadoconliqui().list() / client.Contadoconliqui().load({"id": ...})."""
        from entity.contadoconliqui_entity import ContadoconliquiEntity
        return ContadoconliquiEntity(self, data)


    def CotizacionAmbito(self, data=None) -> "CotizacionAmbitoEntity":
        """Entity factory: client.CotizacionAmbito().list() / client.CotizacionAmbito().load({"id": ...})."""
        from entity.cotizacion_ambito_entity import CotizacionAmbitoEntity
        return CotizacionAmbitoEntity(self, data)


    def Cotizacione(self, data=None) -> "CotizacioneEntity":
        """Entity factory: client.Cotizacione().list() / client.Cotizacione().load({"id": ...})."""
        from entity.cotizacione_entity import CotizacioneEntity
        return CotizacioneEntity(self, data)


    def Cripto(self, data=None) -> "CriptoEntity":
        """Entity factory: client.Cripto().list() / client.Cripto().load({"id": ...})."""
        from entity.cripto_entity import CriptoEntity
        return CriptoEntity(self, data)


    def Dolare(self, data=None) -> "DolareEntity":
        """Entity factory: client.Dolare().list() / client.Dolare().load({"id": ...})."""
        from entity.dolare_entity import DolareEntity
        return DolareEntity(self, data)


    def Estado(self, data=None) -> "EstadoEntity":
        """Entity factory: client.Estado().list() / client.Estado().load({"id": ...})."""
        from entity.estado_entity import EstadoEntity
        return EstadoEntity(self, data)


    def Eur(self, data=None) -> "EurEntity":
        """Entity factory: client.Eur().list() / client.Eur().load({"id": ...})."""
        from entity.eur_entity import EurEntity
        return EurEntity(self, data)


    def Mayorista(self, data=None) -> "MayoristaEntity":
        """Entity factory: client.Mayorista().list() / client.Mayorista().load({"id": ...})."""
        from entity.mayorista_entity import MayoristaEntity
        return MayoristaEntity(self, data)


    def Oficial(self, data=None) -> "OficialEntity":
        """Entity factory: client.Oficial().list() / client.Oficial().load({"id": ...})."""
        from entity.oficial_entity import OficialEntity
        return OficialEntity(self, data)


    def Tarjeta(self, data=None) -> "TarjetaEntity":
        """Entity factory: client.Tarjeta().list() / client.Tarjeta().load({"id": ...})."""
        from entity.tarjeta_entity import TarjetaEntity
        return TarjetaEntity(self, data)


    def Uyu(self, data=None) -> "UyuEntity":
        """Entity factory: client.Uyu().list() / client.Uyu().load({"id": ...})."""
        from entity.uyu_entity import UyuEntity
        return UyuEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "DolarYMonedasSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.blue_entity import BlueEntity
    from entity.bolsa_entity import BolsaEntity
    from entity.brl_entity import BrlEntity
    from entity.clp_entity import ClpEntity
    from entity.contadoconliqui_entity import ContadoconliquiEntity
    from entity.cotizacion_ambito_entity import CotizacionAmbitoEntity
    from entity.cotizacione_entity import CotizacioneEntity
    from entity.cripto_entity import CriptoEntity
    from entity.dolare_entity import DolareEntity
    from entity.estado_entity import EstadoEntity
    from entity.eur_entity import EurEntity
    from entity.mayorista_entity import MayoristaEntity
    from entity.oficial_entity import OficialEntity
    from entity.tarjeta_entity import TarjetaEntity
    from entity.uyu_entity import UyuEntity

File: lua/dolar-y-monedas_sdk.lua
-- DolarYMonedas SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local DolarYMonedasSDK = {}
DolarYMonedasSDK.__index = DolarYMonedasSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

DolarYMonedasSDK._make_feature = _make_feature


function DolarYMonedasSDK.new(options)
  local self = setmetatable({}, DolarYMonedasSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function DolarYMonedasSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function DolarYMonedasSDK:get_utility()
  return Utility.copy(self._utility)
end


function DolarYMonedasSDK:get_root_ctx()
  return self._rootctx
end


function DolarYMonedasSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function DolarYMonedasSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:blue():list() / client:blue():load({ id = ... })
function DolarYMonedasSDK:blue(data)
  local EntityMod = require("entity.blue_entity")
  if data == nil then
    if self._blue == nil then
      self._blue = EntityMod.new(self, nil)
    end
    return self._blue
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:blue() instead.
function DolarYMonedasSDK:Blue(data)
  local EntityMod = require("entity.blue_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:bolsa():list() / client:bolsa():load({ id = ... })
function DolarYMonedasSDK:bolsa(data)
  local EntityMod = require("entity.bolsa_entity")
  if data == nil then
    if self._bolsa == nil then
      self._bolsa = EntityMod.new(self, nil)
    end
    return self._bolsa
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:bolsa() instead.
function DolarYMonedasSDK:Bolsa(data)
  local EntityMod = require("entity.bolsa_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:brl():list() / client:brl():load({ id = ... })
function DolarYMonedasSDK:brl(data)
  local EntityMod = require("entity.brl_entity")
  if data == nil then
    if self._brl == nil then
      self._brl = EntityMod.new(self, nil)
    end
    return self._brl
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:brl() instead.
function DolarYMonedasSDK:Brl(data)
  local EntityMod = require("entity.brl_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:clp():list() / client:clp():load({ id = ... })
function DolarYMonedasSDK:clp(data)
  local EntityMod = require("entity.clp_entity")
  if data == nil then
    if self._clp == nil then
      self._clp = EntityMod.new(self, nil)
    end
    return self._clp
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:clp() instead.
function DolarYMonedasSDK:Clp(data)
  local EntityMod = require("entity.clp_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:contadoconliqui():list() / client:contadoconliqui():load({ id = ... })
function DolarYMonedasSDK:contadoconliqui(data)
  local EntityMod = require("entity.contadoconliqui_entity")
  if data == nil then
    if self._contadoconliqui == nil then
      self._contadoconliqui = EntityMod.new(self, nil)
    end
    return self._contadoconliqui
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:contadoconliqui() instead.
function DolarYMonedasSDK:Contadoconliqui(data)
  local EntityMod = require("entity.contadoconliqui_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:cotizacion_ambito():list() / client:cotizacion_ambito():load({ id = ... })
function DolarYMonedasSDK:cotizacion_ambito(data)
  local EntityMod = require("entity.cotizacion_ambito_entity")
  if data == nil then
    if self._cotizacion_ambito == nil then
      self._cotizacion_ambito = EntityMod.new(self, nil)
    end
    return self._cotizacion_ambito
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:cotizacion_ambito() instead.
function DolarYMonedasSDK:CotizacionAmbito(data)
  local EntityMod = require("entity.cotizacion_ambito_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:cotizacione():list() / client:cotizacione():load({ id = ... })
function DolarYMonedasSDK:cotizacione(data)
  local EntityMod = require("entity.cotizacione_entity")
  if data == nil then
    if self._cotizacione == nil then
      self._cotizacione = EntityMod.new(self, nil)
    end
    return self._cotizacione
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:cotizacione() instead.
function DolarYMonedasSDK:Cotizacione(data)
  local EntityMod = require("entity.cotizacione_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:cripto():list() / client:cripto():load({ id = ... })
function DolarYMonedasSDK:cripto(data)
  local EntityMod = require("entity.cripto_entity")
  if data == nil then
    if self._cripto == nil then
      self._cripto = EntityMod.new(self, nil)
    end
    return self._cripto
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:cripto() instead.
function DolarYMonedasSDK:Cripto(data)
  local EntityMod = require("entity.cripto_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:dolare():list() / client:dolare():load({ id = ... })
function DolarYMonedasSDK:dolare(data)
  local EntityMod = require("entity.dolare_entity")
  if data == nil then
    if self._dolare == nil then
      self._dolare = EntityMod.new(self, nil)
    end
    return self._dolare
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:dolare() instead.
function DolarYMonedasSDK:Dolare(data)
  local EntityMod = require("entity.dolare_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:estado():list() / client:estado():load({ id = ... })
function DolarYMonedasSDK:estado(data)
  local EntityMod = require("entity.estado_entity")
  if data == nil then
    if self._estado == nil then
      self._estado = EntityMod.new(self, nil)
    end
    return self._estado
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:estado() instead.
function DolarYMonedasSDK:Estado(data)
  local EntityMod = require("entity.estado_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:eur():list() / client:eur():load({ id = ... })
function DolarYMonedasSDK:eur(data)
  local EntityMod = require("entity.eur_entity")
  if data == nil then
    if self._eur == nil then
      self._eur = EntityMod.new(self, nil)
    end
    return self._eur
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:eur() instead.
function DolarYMonedasSDK:Eur(data)
  local EntityMod = require("entity.eur_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:mayorista():list() / client:mayorista():load({ id = ... })
function DolarYMonedasSDK:mayorista(data)
  local EntityMod = require("entity.mayorista_entity")
  if data == nil then
    if self._mayorista == nil then
      self._mayorista = EntityMod.new(self, nil)
    end
    return self._mayorista
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:mayorista() instead.
function DolarYMonedasSDK:Mayorista(data)
  local EntityMod = require("entity.mayorista_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:oficial():list() / client:oficial():load({ id = ... })
function DolarYMonedasSDK:oficial(data)
  local EntityMod = require("entity.oficial_entity")
  if data == nil then
    if self._oficial == nil then
      self._oficial = EntityMod.new(self, nil)
    end
    return self._oficial
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:oficial() instead.
function DolarYMonedasSDK:Oficial(data)
  local EntityMod = require("entity.oficial_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:tarjeta():list() / client:tarjeta():load({ id = ... })
function DolarYMonedasSDK:tarjeta(data)
  local EntityMod = require("entity.tarjeta_entity")
  if data == nil then
    if self._tarjeta == nil then
      self._tarjeta = EntityMod.new(self, nil)
    end
    return self._tarjeta
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:tarjeta() instead.
function DolarYMonedasSDK:Tarjeta(data)
  local EntityMod = require("entity.tarjeta_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:uyu():list() / client:uyu():load({ id = ... })
function DolarYMonedasSDK:uyu(data)
  local EntityMod = require("entity.uyu_entity")
  if data == nil then
    if self._uyu == nil then
      self._uyu = EntityMod.new(self, nil)
    end
    return self._uyu
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:uyu() instead.
function DolarYMonedasSDK:Uyu(data)
  local EntityMod = require("entity.uyu_entity")
  return EntityMod.new(self, data)
end




function DolarYMonedasSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = DolarYMonedasSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return DolarYMonedasSDK
